moveit_planners/trajopt/src/trajopt_interface.cpp

```diff
@@ -0,0 +1,452 @@
+
+    // TODO: Feed cart_goal_pos with request information and the needed param to the setup.yaml file
+    // TODO: Multiple Cartesian constraints
+
+    // Add the constraint to problem_info
+    problem_info.cnt_infos.push_back(cart_goal_pos);
+  }
+  else if (req.goal_constraints[0].position_constraints.empty() &&
+           !req.goal_constraints[0].orientation_constraints.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "position constraint is not defined");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_GOAL_CONSTRAINTS;
+    return false;
+  }
+  else if (!req.goal_constraints[0].orientation_constraints.empty() &&
+           req.goal_constraints[0].orientation_constraints.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "orientation constraint is not defined");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_GOAL_CONSTRAINTS;
+    return false;
+  }
+
+  ROS_INFO(" ======================================= Constraints from request goal_constraints");
+  for (auto goal_cnt : req.goal_constraints)
+  {
+    JointPoseTermInfoPtr joint_pos_term(new JointPoseTermInfo);
+    // When using MotionPlanning Display in RViz, the created request has no name for the constriant
+    setJointPoseTermInfoParams(joint_pos_term, (goal_cnt.name != "") ? goal_cnt.name : "goal_tmp" );
+
+    trajopt::DblVec joint_goal_constraints;
+    for (const moveit_msgs::JointConstraint& joint_goal_constraint : goal_cnt.joint_constraints)
+    {
+      joint_goal_constraints.push_back(joint_goal_constraint.position);
+    }
+    joint_pos_term->targets = joint_goal_constraints;
+    problem_info.cnt_infos.push_back(joint_pos_term);
+  }
+
+  ROS_INFO(" ======================================= Constraints from request start_state");
+  // add the start pos from request as a constraint
+  std::unordered_map<std::string, double> all_joints;
+  JointPoseTermInfoPtr joint_start_pos(new JointPoseTermInfo);
+  trajopt::DblVec joint_start_constraints;
+
+  for (int joint_index = 0; joint_index < req.start_state.joint_state.position.size(); ++joint_index)
+  {
+    all_joints[req.start_state.joint_state.name[joint_index]] = req.start_state.joint_state.position[joint_index];
+  }
+
+  for (auto joint_name : joint_names)
+  {
+    ROS_INFO(" ======================================= joint position from start state, name: %s, value: %f", joint_name.c_str(), all_joints[joint_name]);
+    joint_start_constraints.push_back(all_joints[joint_name]);
+  }
+  joint_start_pos->targets = joint_start_constraints;
+  setJointPoseTermInfoParams(joint_start_pos, "start_pos");
+  problem_info.cnt_infos.push_back(joint_start_pos);
+
+  ROS_INFO(" ======================================= Velocity Constraints, hard-coded");
+  // TODO: should be defined by user, its parametes should be added to setup.yaml
+  JointVelTermInfoPtr joint_vel(new JointVelTermInfo);
+
+  joint_vel->coeffs = std::vector<double>(7, 5.0);
```

Comment:
The number of coefficients should match the DOF of the specified joint model group.

Reply:
right, corrected it